### ISSUE TITLE
Automatic refresh in development sometimes doesn't work for simple apps

### ISSUE THREAD

**Issue by astegmaier** [parcel-bundler/parcel#7028]
# 🐛 bug report

Normally, parcel will automatically reload the browser when changes made to html files while running in development mode.

I found a very simple example that breaks this behavior - e.g. with the configuration below, when you make changes to `index.html`, the browser is not automatically refreshed.

## 🎛 Configuration (.babelrc, package.json, cli command)

**src/index.html**
```html
<html>
<head>
  <script src="showAlert.js"></script>
</head>
<body>
  <button onClick="showAlert()">Show Alert</button>
</body>
</html>
```

**src/showAlert.js**
```js
function showAlert() {
  alert("Here is a message from the showAlert function!");
}
```

**package.json**
```json
{
  "scripts": {
    "start": "parcel src/index.html",
  },
  "devDependencies": {
    "parcel": "2.0.0-nightly.859",
  }
}

```
## 🤔 Expected Behavior

If you run `yarn start`, load the page in the browser, then make a change to `index.html`, the page should refresh.

## 😯 Current Behavior

The page doesn't refresh after making changes.

If you remove the `<script src="showAlert.js"></script>` line from `index.html` and refresh the browser, subsequent changes to `index.html` _will_ trigger refresh, as expected.

## 💁 Possible Solution

I noticed that the HMR runtime is not loaded at all in this case - I'm not sure why. One possibility is that one of the conditions in this block is returning "true", which prevents the runtime from loading, but I haven't been able to debug yet:

https://github.com/parcel-bundler/parcel/blob/fe41fc1a01b6f56cedebead424c01288fcf7deb3/packages/runtimes/hmr/src/HMRRuntime.js#L14-L22

## 🔦 Context

I was trying to help someone with [a StackOverflow question](https://stackoverflow.com/questions/69442728/hot-module-replacement-not-working-in-parcel-bundler). It looks like he has a relatively simple, multi-page vanilla-js project. I tend not to build things that way, but I suppose some folks still do, and it would be great if they had a good experience with parcel.

## 💻 Code Sample

You can see [an example of the bug in this repo](https://github.com/astegmaier/parcel-refresh-bug).

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.0-nightly.859
| Node             | 14.17.5
| npm/Yarn         | yarn 1.22.13
| Operating System | Mac OS 11.6 Big Sur

**Comment by mischnic** [comment]
> bundle.env.sourceType === 'script' 

It should work if you add `type="module"`

**Comment by devongovett** [comment]
Indeed. HMR cannot be done correctly for classic scripts because they run in the global scope (i.e. not inside a function). However, we could detect this case in the HTML transformer like we do for HTML files without any scripts at all. This adds an additional script to the page with the HMR runtime, which should cause the page to at least be fully reloaded.

**Comment by astegmaier** [comment]
> It should work if you add type="module"

That works to fix refresh, but it breaks the app. `showAlert.js` is not a module - it's intended to define a global variable so that the function defined there can be bound directly to the button `onClick` property in the html.

(I get that this isn't a great way to build apps, but I suspect it was what the person on StackOverflow may have been doing).

> HMR cannot be done correctly for classic scripts because they run in the global scope (i.e. not inside a function).

I don't think that the user's expectation in this case is that we do proper "HMR" where we hot-swap portions of the page without triggering a full refresh. He would have been happy with triggering a full-page reload.

> However, we could detect this case in the HTML transformer like we do for HTML files without any scripts at all.

I think this change would make this particular user happy. (Not saying it's the highest priority though). If I wanted to make a PR, do you have any pointers?

**Comment by mischnic** [comment]
This is the code that Devon was referring to:

https://github.com/parcel-bundler/parcel/blob/fe41fc1a01b6f56cedebead424c01288fcf7deb3/packages/transformers/html/src/HTMLTransformer.js#L47-L65

**Comment by astegmaier** [comment]
> This is the code that Devon was referring to:

Ah thanks! I think I'll work on a PR when I have time. I don't want to break anything that currently works (e.g. cause a full page refresh where today we're doing proper HMR, or cause the HMR runtime to be included twice), but I suppose we can talk through the details after I have a chance to test things out.

**Comment by astegmaier** [comment]
I experimented with adding a condition to line 47 of `HTMLTransformer.js` so that the top-level HMR runtime is added in cases where there's only global scripts. It works to get the `html` page reloading on changes to it. But when you change a global javascript file referenced by that html page, there's no auto-reload.

I think that's because of this line in the HMR runtime...

https://github.com/parcel-bundler/parcel/blob/daf2cd9c0c53caaef48971e71475957bb5e4e328/packages/runtimes/hmr/src/loaders/hmr-runtime.js#L102

...which won't process changes from websocket messages for assets with a different `bundle.env.id`. I tried adding some logging, and I found that `html` files will share the same `bundle.env.id` with any `<script type="module">` assets that they reference, but not with any `<script>` (no-module) assets they reference.

@devongovett and @mischnic - I don't fully understand the purpose of `bundle.env.id` - _should_ it be different for no-module js files? If not, that might be the fix. If yes, do you have any ideas about the right design?

**Comment by ronnyhaase** [comment]
Any updates here? I was surprised to notice the same problem with a simple vanilla Pug/SCSS/JS setup.
I think this is far from ideal in terms of DX for new joiners... would be cool if Parcel was so smart to do a full-page reload.

**Update:** In my case my `script` Tag was missing `type="module"` and it's properly highlighted in the Get Started guide that it should be added, so I guess my case is obsolete.

**Comment by RyanPridgeon** [comment]
HMR seems to constantly randomly break for me too on Parcel 2.2.1. Some combination of rebooting my computer, deleting .parcel-cache or deleting dist is needed. This happens almost every day.

**Comment by tpudel** [comment]
> 

Every time I change branches I have to do this as well now, keeps old built cached code when it shouldnt! Any ideas?
